File: C11/백준/Silver/1697. 숨바꼭질/숨바꼭질.c
#include <stdio.h>
#include <stdbool.h>

#define MAX 100001

int queue[MAX];
int time[MAX];
bool visited[MAX];
int front = 0, rear = 0;

void enqueue(int x) {
    queue[rear++] = x;
}

int dequeue() {
    return queue[front++];
}

int bfs(int start, int target) {
    enqueue(start);
    visited[start] = true;
    time[start] = 0;

    while (front < rear) {
        int now = dequeue();

        if (now == target) return time[now];

        int next[3] = {now - 1, now + 1, now * 2};
        for (int i = 0; i < 3; i++) {
            int nx = next[i];
            if (nx >= 0 && nx < MAX && !visited[nx]) {
                enqueue(nx);
                visited[nx] = true;
                time[nx] = time[now] + 1;
            }
        }
    }
    return -1;
}

int main() {
    int N, K;
    scanf("%d %d", &N, &K);

    int result = bfs(N, K);
    printf("%d\n", result);

    return 0;
}

File: Java/백준/Silver/1697. 숨바꼭질/숨바꼭질.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {
    static final int MAX = 100001;
    static boolean[] visited = new boolean[MAX];

    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        String[] input = br.readLine().split(" ");
        int N = Integer.parseInt(input[0]);
        int K = Integer.parseInt(input[1]);

        System.out.print(bfs(N, K));
    }

    static int bfs(int start, int target) {
        Queue<int[]> queue = new LinkedList<>();
        queue.add(new int[]{start, 0});
        visited[start] = true;

        while (!queue.isEmpty()) {
            int[] current = queue.poll();
            int position = current[0];
            int time = current[1];

            if (position == target) return time;

            int[] nextPositions = {position - 1, position + 1, position * 2};
            for (int next : nextPositions) {
                if (next >= 0 && next < MAX && !visited[next]) {
                    visited[next] = true;
                    queue.add(new int[]{next, time + 1});
                }
            }
        }

        return -1;
    }
}